### DISCUSSION TITLE
Exception returned from Remote DataPortal does not contain inner exceptions

### DISCUSSION THREAD

**Discussion by GreatBarrier86** [MarimerLLC/csla#1712] [Questions]
Hello,

In 4.9.0, it appears that when the `HttpResponse.ErrorData` is set into the new `DataPortalException` on Line 274 of the `HttpProxy` class, the `HttpErrorInfo.InnerError` property is ignored. Therefore, when the exception thrown from the server ends up containing a `InnerException`, the `ErrorInfo.InnerError` property received by the client is not set into `DataPortalException.InnerException` so we're not able to see what really failed.

Am I doing something wrong? I've traced into the code but I don't see anything that would change this setting.

**HttpProxy Fetch method**
```
        else if (response != null && response.ErrorData != null)
        {
          var ex = new DataPortalException(response.ErrorData);
          result = new DataPortalResult(null, ex, globalContext);
        }
```

**DataPortalException constructor**
```
    public DataPortalException(Csla.Server.Hosts.HttpChannel.HttpErrorInfo info)
      : base(info.Message)
    {
      this.ErrorInfo = info;
    }
```

Since `DataPortalResult` has no direct reference to `response.ErrorData` and it appears that all `new DataPortalException()` does is set the `info.Message` to the base class, the `InnerError` data does not appear accessible.

I would expect some sort of recursive implementation to set the `InnerException` properties all the way down, but 
`Exception.InnerException` cannot be set outside the constructor and I'm not seeing where that would be implemented.

This means that in the catch block of one our BO's DataPortal_Fetch method's, the if statement no longer works
```
        public CustomerMarketing GetPreviousCustomerMarketing()
        {
            try
            {
                return AsyncUtil.RunSync(() => { return DataPortal.FetchAsync<CustomerMarketing>(myCriteria); });
            }
            catch (DataPortalException e)
            {
                if (e.BusinessException is ObjectNotFoundException)
                {
                    return null;
                }
                throw;
            }
        }
```

Thanks,
Jason

**Comment by rockfordlhotka** [comment]
I think that this was address by [this specific change](https://github.com/MarimerLLC/csla/commit/8cf85816829a5221be82dd76e1c01f217098b093#diff-71e583f3445200f1e3f15a03ccea3b1a).

So I am pretty sure the problem is resolved in current versions of the framework.

**Reply by GreatBarrier86** [reply]
@rockfordlhotka, i'd think that if anything, it would be [this](https://github.com/MarimerLLC/csla/commit/6ea02b8624c19b132e2a7437a41364ba2ac43006) change. The one you linked only shows you throwing an `HttpRequestException` and setting a composite `String` value for the message.

That being said, i looked at the link i added and i still don't see where the `HttpErrorInfo.InnerError` is being set as `InnerException`. I can use `HttpErrorInfo.ExceptionTypeName` to compare based on type, but it still seems we've lost the way to look at the `BusinessException` inside the `DataPortalException`

**This is the constructor used by HttpProxy. The base() call is not setting InnerException and InnerException has no setter**
```
    public DataPortalException(Csla.Server.Hosts.HttpChannel.HttpErrorInfo info)
      : base(info.Message)
    {
      this.ErrorInfo = info;
    }
```

**This is the BusinessException property in DataPortalException. It has no setter so it cannot be set from elsewhere**

```
    public Exception BusinessException
    {
      get
      {
        var result = this.InnerException;
        if (result != null)
          result = result.InnerException;
        if (result is DataPortalException dpe && dpe.InnerException != null)
          result = dpe.InnerException;
        if (result is Csla.Reflection.CallMethodException cme && cme.InnerException != null)
          result = cme.InnerException;
        return result;
      }
    }
```

So I guess i cannot see how `DataPortalException.InnerException` would be set from this constructor, which is the constructor used by `HttpProxy.
`

**Reply by rockfordlhotka** [reply]
I'm not sure what you are expecting?

Actual _exception_ types/objects can not flow over the wire. It was never really reliable, even with BinaryFormatter, and with MobileFormatter it isn't possible at all.

The server-side exception graph (including inner exceptions) should be used to create an ErrorInfo graph that mirrors the original exception graph.

If you want all the details about the server exceptions, you get them by walking the ErrorInfo graph.

**Reply by GreatBarrier86** [reply]
It's very possible what i'm expecting isn't doable.

**Simply put, the result is different between the Local and Remote. I expected Remote would be able to have an InnerException specified**
1. LocalProxy: `DataPortalException` has an `InnerException` specified
2. RemoteProxy: `DataPortalException` has an `InnerException` set to NULL

I guess it's really not possible though since Exceptions don't have properties with public setters, so you can't really create new ones and deserialize them as normal.

Does what i'm saying at least make sense? :)

**Reply by rockfordlhotka** [reply]
It does make sense. I think we could make the local data portal act like the remote data portal, but we can't make the remote act like local.
